### PR TITLE
[Profile] Applying format.ps1xml generation mechanism [Iss #6183]

### PR DESCRIPTION
**Pull request by vladimir-shcherbakov**
## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [ ] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [ ] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [ ] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by vladimir-shcherbakov** [comment]
sign build is here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/643/

File: documentation/development-docs/generate-format.ps1xml-file.md [hunk 1]
```diff
@@ -136,12 +136,20 @@ namespace Microsoft.WindowsAzure.Commands.Common.Attributes
...
 
         public ViewControl Target { get; set; } = ViewControl.Table;
+
+        public string ScriptBlock { get; set; }
+
+        public bool GroupByThis { get; set; }
+
+        public uint TableColumnWidth { get; set; }
+
+        public uint Position { get; set; } = Ps1XmlConstants.DefaultPosition;
```
Comment (praries880) [inline documentation/development-docs/generate-format.ps1xml-file.md]: so, what happens if :
* A user specified the same position on multiple properties in the same class?

* Not all properties in the class have a position specified?

File: documentation/development-docs/generate-format.ps1xml-file.md [hunk 2]
```diff
@@ -136,12 +136,20 @@ namespace Microsoft.WindowsAzure.Commands.Common.Attributes
...
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field, AllowMultiple = true)]
     public sealed class Ps1XmlAttribute : Attribute
     {
         public string Label { get; set; }
 
         public ViewControl Target { get; set; } = ViewControl.Table;
+
+        public string ScriptBlock { get; set; }
+
+        public bool GroupByThis { get; set; }
```
Comment (praries880) [inline documentation/development-docs/generate-format.ps1xml-file.md]: what happens if multiple properties in the same class?

File: src/Common/Commands.Common/Attributes/Ps1XmlAttribute.cs [hunk 3]
```diff
@@ -23,11 +23,24 @@ public enum ViewControl
...
 
         public ViewControl Target { get; set; } = ViewControl.Table;
+
+        public string ScriptBlock { get; set; }
+
+        public bool GroupByThis { get; set; }
+
+        public uint TableColumnWidth { get; set; }
+
+        public uint Position { get; set; } = Ps1XmlConstants.DefaultPosition;
```
Comment (praries880) [inline src/Common/Commands.Common/Attributes/Ps1XmlAttribute.cs]: so, what happens if :
* A user specified the same position on multiple properties in the same class?

* Not all properties in the class have a position specified?

File: src/Common/Commands.Common/Attributes/Ps1XmlAttribute.cs [hunk 4]
```diff
@@ -23,11 +23,24 @@ public enum ViewControl
...
+    [AttributeUsage(AttributeTargets.Property | AttributeTargets.Field, AllowMultiple = true)]
     public sealed class Ps1XmlAttribute : Attribute
     {
         public string Label { get; set; }
 
         public ViewControl Target { get; set; } = ViewControl.Table;
+
+        public string ScriptBlock { get; set; }
+
+        public bool GroupByThis { get; set; }
```
Comment (praries880) [inline src/Common/Commands.Common/Attributes/Ps1XmlAttribute.cs]: What is multiple properties in the class have this true?

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs [hunk 5]
```diff
@@ -161,61 +196,164 @@ internal Configuration BuildXmlConfiguration(IEnumerable<Type> cmdletTypList)
...
+                                    PropertyName = memberInfo.Name,
+                                    TargetView = ps1XmlAttribute.Target,
+                                };
+
+                                continue;
+                            }
+                            else
+                            {
+                                // duplicate!
+                                // FIXME: replace or ignore or throw?
```
Comment (praries880) [inline tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: whats going to happen in this case? currently we seem to be only using the first group by we see.. the later ones will be ignored. We should leave th e user with a good default behavior.
I think we can even throw an exception and say that only one Group By is allowed.

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs [hunk 6]
```diff
@@ -161,61 +196,164 @@ internal Configuration BuildXmlConfiguration(IEnumerable<Type> cmdletTypList)
...
+                                };
+
+                            if (position == Ps1XmlConstants.DefaultPosition)
+                            {
+                                tableHeaders.Add(tableColumnHeader);
+                                tableColumnItems.Add(tableColumnItem);
+                            }
+                            else
+                            {
+                                columnHeadersWithPosition.Add(Tuple.Create(tableColumnHeader, position));
```
Comment (praries880) [inline tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: what is multiple entries come up with the same position? Should we throw an exception saying that multiple entries found with the same position?

Also, what if the user starts the position at a non zero number? or does not specify position on all attributes?

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs [hunk 7]
```diff
@@ -161,61 +197,205 @@ internal Configuration BuildXmlConfiguration(IEnumerable<Type> cmdletTypList)
...
             if (memberInfoListFiltered.Count == 0) return;
 
+            var columnHeadersWithPosition = new List<Tuple<TableColumnHeader, uint>>();
+            var columnItemsWithPosition = new List<Tuple<TableColumnItem, uint>>();
+            var listItemsWithPosition = new List<Tuple<ListItem, uint>>();
+
+            var dupTrackerGroupBy = new List<Tuple<string, string, string>>();
+            var dupTrackerPosition = new List<Tuple<string, string, string, uint>>();
+
             foreach (var memberInfo in memberInfoListFiltered)
```
Comment (praries880) [inline tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: It'll be great if you can add a comment here that explains whats happening in the for loop and how we track duplicates for both position and group by.

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs [hunk 8]
```diff
@@ -161,61 +197,205 @@ internal Configuration BuildXmlConfiguration(IEnumerable<Type> cmdletTypList)
...
                     {
                         PropertyName = memberInfo.Name
                     });
                 }
             }
+
+            // track duplicate
+            if (dupTrackerGroupBy.Count > 1)
+            {
+                var warning = "Found a duplicate of property'GroupByThis'. 'GroupByThis=true' have the following properties: ";
```
Comment (praries880) [inline tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: Nitpick : a space is needed between `property` and `GroupByThis`

What do you think of the wording :  _"`GroupByThis` is `true` for the following properties"_

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs [hunk 9]
```diff
@@ -161,61 +197,205 @@ internal Configuration BuildXmlConfiguration(IEnumerable<Type> cmdletTypList)
...
+            }
+
+            var duplicates = dupTrackerPosition
+                .GroupBy(p => p.Item4)
+                .Where(g => g.Count() > 1)
+                .ToList();
+
+            if (duplicates.Count > 0)
+            {
+                var warning = "Found a duplicate of property'Position'. ";
```
Comment (praries880) [inline tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: Nitpick : a space is needed between `property` and `Position`

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs [hunk 10]
```diff
@@ -161,61 +197,205 @@ internal Configuration BuildXmlConfiguration(IEnumerable<Type> cmdletTypList)
...
+                foreach (var duplicate in duplicates)
+                {
+                    warning += $"'Positon={duplicate.Key}' have the following properties:";
```
Comment (praries880) [inline tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: Nitpick : What do you think of the wording :  _Following properties had the same `Position` value of `{duplicate.key} specified:`"_

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs (hunk 9 above)
Comment (vladimir-shcherbakov) [reply tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: fixed

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs (hunk 10 above)
Comment (vladimir-shcherbakov) [reply tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: fixed

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs (hunk 8 above)
Comment (vladimir-shcherbakov) [reply tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: fixed

File: tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs [hunk 11]
```diff
@@ -161,61 +197,205 @@ internal Configuration BuildXmlConfiguration(IEnumerable<Type> cmdletTypList)
...
-                    (ps1XmlAttribute.Target & ViewControl.List) != ViewControl.None)
-                {
                     listItems.Add(new ListItem
                     {
                         PropertyName = memberInfo.Name
                     });
                 }
             }
+
+            // track duplicate
```
[the remark this answers is not in the document]
Comment (vladimir-shcherbakov) [reply tools/RepoTasks/RepoTasks.RemoteWorker/AppDomainWorker.cs]: fixed

**Comment by praries880** [comment]
@vladimir-shcherbakov  Changes look good to me. Will merge this once the release branch is merged back in preview branch (sometime early next week).

**Comment by vladimir-shcherbakov** [comment]
on demand: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/945/
sign: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/770/

**Comment by cormacpayne** [comment]
@vladimir-shcherbakov Can you please fix the merge conflict? We have moved ChangeLog/psd1 files from `src/ResourceManager/` to `src/ResourceManager/{{service}}/Commands.{{service}}`, so you will need to move your ChangeLog/psd1 updates there.

**Comment by vladimir-shcherbakov** [comment]
on demand: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/947/
sign: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/ps-sign/774/

**Pull request merged**